pkg/tsdb/prometheus/json_to_dataframes_test.go

```diff
@@ -0,0 +1,258 @@
+		return backend.DataResponse{}, fmt.Errorf("result must contain only one DataResponse, but contained: %v", getRefs(result))
+	}
+
+	dr, found := result.Responses["A"]
+
+	if !found {
+		return backend.DataResponse{}, fmt.Errorf("DataResponse not found in result, it contained: %v", getRefs(result))
+	}
+
+	return dr, nil
+}
+
+func TestRunQuery(t *testing.T) {
+	t.Run("parse a simple matrix response", func(t *testing.T) {
+		// NOTE: the time-related fields in the query structure
+		// have to match with the data in the JSON_response,
+		// because the code auto-generates missing values.
+		start := time.Unix(1641889530, 0)
+
+		query := PrometheusQuery{
+			RefId:      "A",
+			RangeQuery: true,
+			Start:      start,
+			End:        start.Add(time.Second * 2),
+			Step:       time.Second,
+		}
+
+		response := `
+		{
+			"status": "success",
+			"data": {
+				"resultType": "matrix",
+				"result": [
+					{
+						"metric": { "__name__": "go_goroutines", "job": "prometheus" },
+						"values": [
+							[1641889530, "21"],
+							[1641889531, "32"],
+							[1641889532, "43"]
+						]
+					}
+				]
+			}
+		}
+
+		`
+
+		dr, err := getResponse([]byte(response), query)
+		require.NoError(t, err)
+		require.NoError(t, dr.Error)
+
+		err = experimental.CheckGoldenDataResponse(filepath.Join("json_to_dataframes_testdata", "range_simple.golden.txt"), &dr, true)
+		require.NoError(t, err)
+	})
+
+	t.Run("parse a simple matrix response with value missing steps", func(t *testing.T) {
+		// NOTE: the time-related fields in the query structure
+		// have to match with the data in the JSON_response,
+		// because the code auto-generates missing values.
+		start := time.Unix(1641889530, 0)
+
+		query := PrometheusQuery{
+			RefId:      "A",
+			RangeQuery: true,
+			Start:      start,
+			End:        start.Add(time.Second * 8),
+			Step:       time.Second,
+		}
+
+		response := `
+		{
+			"status": "success",
+			"data": {
+				"resultType": "matrix",
+				"result": [
+					{
+						"metric": { "__name__": "go_goroutines", "job": "prometheus" },
+						"values": [
+							[1641889533, "21"],
+							[1641889534, "32"],
+							[1641889537, "43"]
+						]
+					}
+				]
+			}
+		}
+
+		`
+
+		dr, err := getResponse([]byte(response), query)
+		require.NoError(t, err)
+		require.NoError(t, dr.Error)
+
+		err = experimental.CheckGoldenDataResponse(filepath.Join("json_to_dataframes_testdata", "range_missing.golden.txt"), &dr, true)
+		require.NoError(t, err)
+	})
+
+	t.Run("parse a response with Infinity", func(t *testing.T) {
+		// NOTE: the time-related fields in the query structure
+		// have to match with the data in the JSON_response,
+		// because the code auto-generates missing values.
+		start := time.Unix(1641889530, 0)
+
+		query := PrometheusQuery{
+			RefId:      "A",
+			RangeQuery: true,
+			Start:      start,
+			End:        start.Add(time.Second * 2),
+			Step:       time.Second,
+			Expr:       "1 / 0",
+		}
+
+		response := `
```

Comment:
rather than having the JSON inline in the test files... can we stick them in the `testdata` folder?  I find it much easier to audit and quickly understand what conversions are happening by looking at:

```
testdata/aaa.json
testdata/aaa.golden.txt
testdata/bbb.json
testdata/abb.golden.txt
```

Or maybe:
```
testdata/aaa.query.json
testdata/aaa.result.json
testdata/aaa.result.golden.txt
```

Reply:
In many datasources, I will write a test (that gets skipped) that will load queries and then save the results in files like `aaa.result.json` and then the unit tests are checking result <> golden conversions

Reply:
will do. i'm going to go with the second approach, because the query-params are important while processing the prometheus-response-json (because null-timestamps are inserted based on start-end-timestamp & step)